test/ServiceTest.cpp

```diff
@@ -0,0 +1,179 @@
+                                           parsedQuery::GraphPattern{}};
+
+  // This is the (port-normalized) URL and (whitespace-normalized) SPARQL query
+  // we expect.
+  std::string expectedUrl = "http://localhorst:80/api";
+  std::string expectedSparqlQuery =
+      "PREFIX doof: <http://doof.org> SELECT ?x ?y WHERE { }";
+
+  // CHECK 1: Returned TSV is empty -> an exception should be thrown.
+  Service serviceOperation1{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(expectedUrl, expectedSparqlQuery, "")};
+  ASSERT_THROW(serviceOperation1.getResult(), ad_utility::AbortException);
+
+  // CHECK 2: Header row of returned TSV is wrong (variables in wrong order) ->
+  // an exception should be thrown.
+  Service serviceOperation2{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(
+          expectedUrl, expectedSparqlQuery,
+          "?y\t?x\n<x>\t<y>\n<bla>\t<bli>\n<blu>\t<bla>\n<bli>\t<blu>\n")};
+  ASSERT_THROW(serviceOperation2.getResult(), ad_utility::AbortException);
+
+  // CHECK 3: Returned TSV has correct format matching the query -> check that
+  // the result table returned by the operation corresponds to the contents of
+  // the TSV and its local vocabulary are correct.
+  Service serviceOperation3{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(
+          expectedUrl, expectedSparqlQuery,
+          "?x\t?y\n<x>\t<y>\n<bla>\t<bli>\n<blu>\t<bla>\n<bli>\t<blu>\n")};
+  std::shared_ptr<const ResultTable> result = serviceOperation3.getResult();
+
+  // Check that `<x>` and `<y>` were contained in the original vocabulary and
+  // that `<bla>`, `<bli>`, `<blu>` were added to the (initially empty) local
+  // vocabulary.
+  VocabIndex idxX;
+  VocabIndex idxY;
+  EXPECT_TRUE(testQec->getIndex().getVocab().getId("<x>", &idxX));
+  EXPECT_TRUE(testQec->getIndex().getVocab().getId("<y>", &idxY));
+  LocalVocabIndex idxBla = LocalVocabIndex::make(0);
+  LocalVocabIndex idxBli = LocalVocabIndex::make(1);
+  LocalVocabIndex idxBlu = LocalVocabIndex::make(2);
+  EXPECT_EQ(result->localVocab().getWord(idxBla), "<bla>");
+  EXPECT_EQ(result->localVocab().getWord(idxBli), "<bli>");
+  EXPECT_EQ(result->localVocab().getWord(idxBlu), "<blu>");
+
+  // Check that the result table corresponds to the contents of the TSV.
+  EXPECT_TRUE(result);
+  EXPECT_EQ(result->_idTable.numColumns(), 2);
+  EXPECT_EQ(result->_idTable.numRows(), 4);
+  Id idX = Id::makeFromVocabIndex(idxX);
+  Id idY = Id::makeFromVocabIndex(idxY);
+  Id idBla = Id::makeFromLocalVocabIndex(idxBla);
+  Id idBli = Id::makeFromLocalVocabIndex(idxBli);
+  Id idBlu = Id::makeFromLocalVocabIndex(idxBlu);
+  auto checkRow = [&](size_t rowIdx, Id id1, Id id2) {
+    EXPECT_EQ(result->_idTable(rowIdx, 0), id1) << "at row " << rowIdx;
+    EXPECT_EQ(result->_idTable(rowIdx, 1), id2) << "at row " << rowIdx;
+  };
+  checkRow(0, idX, idY);
+  checkRow(1, idBla, idBli);
+  checkRow(2, idBlu, idBla);
+  checkRow(3, idBli, idBlu);
```

Comment:
You could use the `makeIdTableFromIdVector` from `IdTableHelpers.h`
and then just compare the two `IdTable`s via `operator==`.